package/network/services/dnsmasq/files/dnsmasq.init

```diff
@@ -509,6 +509,11 @@ dhcp_boot_add() {
 	dhcp_option_add "$cfg" "$networkid" "$force"
 }
 
+eval_ipcalc() {
+	local output
```

Comment:
The `eval_` was intented more as a warning about the internals :sweat_smile: 

I'm not entirely convinced of the use of environment variables.
Consider the use of `ipcalc.sh` for isc-dhcpd: If `BROADCAST` was set for a previous UCI section, we don't want that value to leak into subsequent calls of `dhcpd_add`.
For some reason I don't understand, `local v; export -- v=test` doesn't alter any environment variables outside of the function (at least with bash and busybox ash).

There [is a](https://github.com/openwrt/packages/pull/21404/files#diff-ac27427e7c4b64d6e9b62ac528b66be6c626d80ae38ddfad12c6a826a1c5fccdR412) `local .. BROADCAST` before the call to ipcalc. Either `poplocalvars` (seen [here](https://github.com/brgl/busybox/blob/098b713c7b5b22fc60b606be97da431eaedc6639/shell/ash.c#L9145)) restores environment variables or `setvareq` works in such a way that `export` can create an environment variable that is local to the function.

The 'lucky' behaviour of `local` is not documented anywhere that I could find.
It seems a little obscure to me and I prefer the `eval` because I imagine it's easier to understand what's going on.

Reply:
> If BROADCAST was set for a previous UCI section, we don't want that value to leak into subsequent calls of dhcpd_add.
For some reason I don't understand, local v; export -- v=test doesn't alter any environment variables outside of the function (at least with bash and busybox ash).

They’re not “environment” variables if they aren’t exported into another process; they’re just “variables”. Scoping them as `local` stops this from happening.

Reply:
There’s nothing “lucky” about how `local` works. It does exactly what it’s designed to. From `dash(1)`:

>  Functions
>      The syntax of a function definition is
> 
>            name () command
> 
>      A function definition is an executable statement; when executed it
>      installs a function named name and returns an exit status of zero.
>      The command is normally a list enclosed between “{” and “}”.
> 
>      Variables may be declared to be local to a function by using a
>      local command.  This should appear as the first statement of a
>      function, and the syntax is
> 
>            local [variable | -] ...
> 
>      Local is implemented as a builtin command.
> 
>      When a variable is made local, it inherits the initial value and
>      exported and readonly flags from the variable with the same name in
>      the surrounding scope, if there is one.  Otherwise, the variable is
>      initially unset.  The shell uses dynamic scoping, so that if you
>      make the variable x local to function f, which then calls function
>      g, references to the variable x made inside g will refer to the
>      variable x declared inside f, not to the global variable named 

x.

Reply:
> Consider the use of `ipcalc.sh` for isc-dhcpd: If `BROADCAST` was set for a previous UCI section, we don't want that value to leak into subsequent calls of `dhcpd_add`.

I’m the maintainer of `isc-dhcp` and I use stacked (nested) `local`’s without any issues. If need be, we can rename any variables that are problematic.

In any case, isc-dhcp is being retired as soon as I can finish getting the front end to synthesize Kea configuration instead.

Reply:
Lastly, `eval`’s each cost an extra fork/exec pair which is a *lot* slower than staying inside the same process. They’re to be avoided if there’s another way.

Reply:
Ok. My biggest misunderstand was that `export` only marks variables for being exported when a command is run - rather than altering the environment. Insofar, `local` really doesn't do anything special in this case.

But I wasn't arguing the use of `eval`. I was pondering the use of `ip_net="$(ipcalc "$ip/$mask" && echo "$NETWORK")"`
instead of `ipcalc "$ip/$mask" && ip_net="$NETWORK"` because it requires only one `local ip_net` to avoid leaking or possibly overwriting foreign variables.

I hadn't yet considered fork/exec overhead because it's not on a critical code path, at least none that I know of.
Also, there was an `eval` there before, so we're not making anything worse.

Reply:
> I hadn't yet considered fork/exec overhead because this is not a critical code path, at least not to my knowledge.
> Also, there was an eval there before, so we're not making anything worse.

As it happens, the rewrite of the dhcp synthesizer is going to make heavy use of `ipcalc.sh` to figure out which subnet a host belongs in... otherwise you have to use a single global pool and the performance doesn't scale up well.

Reply:
@pprindeville Ok, so there are plans for ipcalc. I can't tell from your description whether that makes it critical for performance (still only used during initialization?).

For 'heavy use', I recommend creating a new interface.
Shell is not well-suited for these 'dictionary' type return values. Best I can think of is handing an environment to a sub process (or to a function call, when using `source`). Ucode, on the other hand, has objects.
@ynezz talked about leveraging ucode's abilities and sharing code with fw4 to extend on [the other PR](https://github.com/openwrt/openwrt/pull/13158).

But is that in the scope of this PR?
Imho, this PR rather needs a decision on which one of these is more appropriate:

```shell
local ip_net
ip_net="$(ipcalc "$ip/$mask" && echo "$NETWORK")"
# or
local ip_net BROADCAST START END NETWORK NETMASK PREFIX
ipcalc "$ip/$mask" && ip_net="$NETWORK"
```

Also, is it worth adding the extra function in `dhcp.script` and `6rd.sh`?
Mt. St. Optimization solves the space elevator problem :stuck_out_tongue: 
I only wanted to fix a crash-loop a year ago.

I'm replacing `ipcalc_eval` with your neat `ipcalc` function for dnsmasq - low hanging fruit (https://github.com/openwrt/openwrt/pull/12925/commits/af098729edaba0d9e6a6a0363c1599fa898be662).
Then, we should only be talking about https://github.com/openwrt/openwrt/pull/12925/commits/7b6d9f98ef2653469f464ad01357c2df430c122f

Reply:
I wish busybox `awk` supported `-E` (I've opened a bug requesting it) so we could do what `ipcalc` on RedHat does (you can pass in flags to only get individual values... in this case, `-n` would limit the output to `NETWORK=`).

I'd go with the 2nd format.

Initialization doesn't always just happen at startup.  Re-initialization might be required if a watched interface bounces, or if the config file gets touched and needs to be reloaded, etc.

Reply:
alright! force pushed.